rest_tools/server/arghandler.py

```diff
@@ -0,0 +1,153 @@
+    ) -> Any:
+        """Cast `value` to `type_` type, and/or check `value` in in `choices`.
+
+        Raise _UnqualifiedArgumentError if either qualification fails.
+        """
+        if type_:
+            try:
+                if isinstance(value, str) and (type_ == bool) and (value != ""):
+                    value = bool(distutils.util.strtobool(value))
+                else:
+                    value = type_(value)
+            except ValueError as e:
+                raise _UnqualifiedArgumentError(f"(ValueError) {e}")
+
+        if choices and (value not in choices):
+            raise _UnqualifiedArgumentError(
+                f"(ValueError) {value} not in options ({choices})"
+            )
+
+        return value
+
+    @staticmethod
+    def _type_check(type_: Optional[type], value: Any) -> None:
+        # check the value's type (None is okay too)
+        if not type_:
+            return
+        if not isinstance(value, type_) and (value is not None):
+            raise ValueError(
+                f"Value, `{value}` ({type(value)}), is not {type_} or None"
+            )
+
+    @staticmethod
+    def get_json_body_argument(  # pylint: disable=R0913
+        request_handler: tornado.web.RequestHandler,
+        name: str,
+        default: Any,
+        choices: Optional[List[Any]],
+    ) -> Any:
+        """Return the argument from JSON-decoded request body."""
+        try:
+            value = _get_json_body(request_handler)[name]
+            return ArgumentHandler._qualify_argument(None, choices, value)
+        except (KeyError, json.decoder.JSONDecodeError):
+            # Required -> raise 400
+            if isinstance(default, type(NO_DEFAULT)):
+                raise _make_400_error(name, tornado.web.MissingArgumentError(name))
+        except _UnqualifiedArgumentError as e:
+            raise _make_400_error(name, e)
+
+        # Else:
+        # Optional / Default
+        try:
+            return ArgumentHandler._qualify_argument(None, choices, default)
+        except _UnqualifiedArgumentError as e:
+            raise _make_400_error(name, e)
+
+    @staticmethod
+    def get_argument(  # pylint: disable=W0221,R0913
+        request_handler: tornado.web.RequestHandler,
+        name: str,
+        default: Any,
+        strip: bool,
+        type_: Optional[type],
+        choices: Optional[List[Any]],
+    ) -> Any:
+        """Return argument from query arguments or JSON request body.
+
+        Try from `get_json_body_argument()` first, then from
+        `request_handler.get_argument()`.
+        """
+        # If:
+        # Required -> raise 400
+        if isinstance(default, type(NO_DEFAULT)):
+            # check JSON-body arguments
+            try:
+                json_arg = ArgumentHandler.get_json_body_argument(
+                    request_handler, name, NO_DEFAULT, choices
+                )
+                return ArgumentHandler._qualify_argument(type_, choices, json_arg)
```

Comment:
Did you consider passing `type_` down into `get_json_body_argument`?

You seem to qualify it there without type (`type_ = None`), which is just a check against the enumerated list `choices` (if provided), then re-qualify it here with the type.

No problem with correctness, just seemed a little odd to my eyes.

Maybe it's a parallel construction with `get_argument` below, so the code here has a similar structure?
get()
qualify()

Reply:
Good eye. I had type-casting in `get_json_body_argument()` several commits back, but I removed it since JSON preserves types (at least built-in ones) so that was moot.

I do type-cast possible JSON-argument values in `get_argument()`.